### PR TITLE
Cookie preferences & End User Agreement per account - initial registries update

### LINKED ISSUE #2807: Cookie Preferences per account
Estimate: none
Expressing interest: none

Store which accounts (Epeople) allow or disallow cookies

For additional details on Cookie Preferences agreement see this comment on the Angular UI ticket: https://github.com/DSpace/dspace-angular/issues/735#issuecomment-670214947

### LINKED ISSUE #2808: User Agreement / Privacy Statement
Estimate from release plan: 24 hours
Expressing interest: 4Science

From release plan: 
To be agreed at self-registration and on login if they got updated

### PR DESCRIPTION
**Pull request by jonas-atmire**
## References
* Fixes [GitHub issue](https://github.com/DSpace/DSpace/issues/2807)

## Description
This adds in the initial metadatafields for the cookie preference related work.
The current fields have been provided by @artlowel as he is handling the Angular version implementation of this.

## Instructions for Reviewers
This basically only adds in some metadatafields, and adds in a check to make sure that these metadatas exist and can be filled in on an eperson level.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

File: dspace/config/registries/dspace-types.xml
Comment (abollini) [inline dspace/config/registries/dspace-types.xml:22]: I would prefer to have `dspace.agreements.acceptedDate` storing the timestamp of the acceptance. Anyway it would be better to add some scope note about how it will be used

Comment (tdonohue) [reply dspace/config/registries/dspace-types.xml:22]: @abollini : I think you might be misunderstanding the point of this metadata field.  There's not just a *single* user/usage agreement, so the field `dspace.agreements.acceptedDate` seems wrong, as it implies there's one agreement with one date of acceptance.

Instead, there's (at least) two different agreements required by the GDPR Working Group:
1. A cookie consent form, described in more detail here: https://github.com/DSpace/dspace-angular/issues/735#issuecomment-670214947
2. An End User Agreement, described in more detail here: https://github.com/DSpace/DSpace/issues/2808#issuecomment-670215129

So, this PR is creating two new fields:
1. `dspace.agreements.cookies` to store an EPerson's answer to the Cookie consent form. My assumption is this field will store the list of cookies that the EPerson has chosen to allow (opt-in approach).  It likely cannot be just a true/false field as we had talked previously about using something like https://klaro.kiprotect.com/#features
2. `dspace.agreements.end-user` to store an EPerson's answer to the End User agreement.  My assumption is that this field _may be_ simply `true` or `false` based on whether they agree with the End User agreement.  However, we could choose to have this field store the date/time of agreement.
    1. Arguably, if this field is ever `false` or `null` then that EPerson account should be blocked until they agree.  It should not be possible to use DSpace if you don't agree to the End User Agreement.

Overall, I agree that we need to add in `<scope_note>` descriptions to both of these fields in order to describe the usage and valid/accepted values.  Currently, while the fields look OK at a glance, I've had to make assumptions on how I expect them to be used (but I'm not sure whether my assumptions are correct).

@pnbecker : You may also want to review the direction of this small PR to provide any feedback on behalf of the GDPR Working Group

Comment (artlowel) [reply dspace/config/registries/dspace-types.xml:22]: >1. `dspace.agreements.cookies` to store an EPerson's answer to the Cookie consent form. My assumption is this field will store the list of cookies that the EPerson has chosen to allow (opt-in approach).  It likely cannot be just a true/false field as we had talked previously about using something like https://klaro.kiprotect.com/#features

That is correct, it will store a copy of the klaro cookie, which is an array of cookieName/boolean pairs. The contents of this metadata field wil be set as the klaro cookie for the session every time you log in, and kept up to date every time you change something in the klaro panel.

>2. `dspace.agreements.end-user` to store an EPerson's answer to the End User agreement.  My assumption is that this field _may be_ simply `true` or `false` based on whether they agree with the End User agreement.  However, we could choose to have this field store the date/time of agreement.
    i. Arguably, if this field is ever `false` or `null` then that EPerson account should be blocked until they agree.  It should not be possible to use DSpace if you don't agree to the End User Agreement.

Correct. If this metadata field has any value other than true, the user will see the end-user agreement immediately after logging in, and can't do anything else until they agree. Disagreeing or canceling will log them out again. If there is an updated version of the user agreement the idea is that you delete this field for all EPersons, so they all have to agree again. We could store a timestamp here instead if that's preferred but I don't see a use case for it at the moment.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@jonas-atmire and @artlowel : Overall, I approve of this PR.  I just have very minor requests inline to better document the usage of these new fields.  I'm OK with the fields as-is, and also OK with the `end-user` field simply having true/false values.  If we discover a later need to store the date of acceptance, that can always be updated at a later time.

File: dspace/config/registries/dspace-types.xml
Comment (jonas-atmire) [reply dspace/config/registries/dspace-types.xml]: Agreed that the scope note needed to be added.
I'd made a mental note to have that updated, but seems that my memory had failed me to actually go through on that as well.
Thanks for the feedback, we've updated the scopes notes for these new fields

File: dspace-api/src/test/java/org/dspace/eperson/EPersonTest.java
Comment (jonas-atmire) [reply dspace-api/src/test/java/org/dspace/eperson/EPersonTest.java]: I'd checked with the Angular team for an example of the klaro cookie, so we've used this as in our test.
I totally agree that actually having relevant information in the test would result in the actual test being more understood.

**Review by tdonohue — APPROVED** [review]
👍 Looks good to me now. Thanks for the minor updates @jonas-atmire !

@pnbecker : Pinging you one last time on this PR to ensure you have a chance to give feedback (if necessary) prior to merger.  I believe this aligns with the goals expressed by the GDPR working group, but would appreciate it if you could verify.  It's a tiny PR, so it should only take you 5-10 mins to glance at it.  _(I plan to merge this on Monday.  Once merged, we can always refactor later on of course.)_

**Pull request merged**
